### PR TITLE
Create backend and dashboard for creating articles

### PR DESCRIPTION
**Pull request by junlarsen**
Does currently not handle tags

### CODE REVIEW

File: packages/types/src/article.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+export const ArticleTagWrite = ArticleTagSchema
+export type ArticleTagWrite = z.infer<typeof ArticleTagWrite>
+
+export const ArticleSchema = z.object({
+  id: z.string().ulid(),
+  createdAt: z.date(),
+  updatedAt: z.date(),
+  title: z.string().min(1),
+  author: z.string().min(1),
+  photographer: z.string().min(1),
```
Comment (junlarsen) [inline packages/types/src/article.ts:19]: Could *technically* have a relation onto OwUser here, but I don't think todays use of this field really warrants it. It also makes the query for fetching the names more complex as we have to go through AWS Cognito

File: packages/db/src/migrations/0020_add_article.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { type Kysely, sql } from "kysely"
+import { createTableWithDefaults } from "../utils"
+
+export async function up(db: Kysely<any>) {
+  await createTableWithDefaults("articles", { id: true, createdAt: true, updatedAt: true }, db.schema)
+    .addColumn("title", "text", (col) => col.notNull())
+    .addColumn("author", "text", (col) => col.notNull())
+    .addColumn("photographer", "text", (col) => col.notNull())
+    .addColumn("image_url", "text", (col) => col.notNull())
```
Comment (henrikskog) [inline packages/db/src/migrations/0020_add_article.ts:9]: Should it be mandatory to upload an image for the article?

File: packages/db/src/migrations/0020_add_article.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import { type Kysely, sql } from "kysely"
+import { createTableWithDefaults } from "../utils"
+
+export async function up(db: Kysely<any>) {
+  await createTableWithDefaults("articles", { id: true, createdAt: true, updatedAt: true }, db.schema)
+    .addColumn("title", "text", (col) => col.notNull())
+    .addColumn("author", "text", (col) => col.notNull())
+    .addColumn("photographer", "text", (col) => col.notNull())
```
Comment (henrikskog) [inline packages/db/src/migrations/0020_add_article.ts:8]: Won't there be situations where there is no obvious photographer for an image? Such as ai generated images etc. Do we want to enforce having a photographer on the db level?

File: packages/db/src/migrations/0020_add_article.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    .addColumn("content", "text", (col) => col.notNull())
+    .execute()
+  await db.schema
+    .createTable("article_tags")
+    .addColumn("name", "text", (col) => col.primaryKey())
+    .execute()
+  await db.schema
+    .createTable("article_tag_link")
+    .addColumn("article", sql`ulid`, (col) => col.references("articles.id").notNull())
+    .addColumn("tag", "text", (col) => col.references("article_tags.name").notNull())
```
Comment (henrikskog) [inline packages/db/src/migrations/0020_add_article.ts:21]: cascade on delete?

File: packages/types/src/article.ts (hunk 1 above)
Comment (henrikskog) [reply packages/types/src/article.ts:19]: In my opinion it's kind of unfortunate to not establish a relation to the given user for the author field. In the future we may want to do something like abakus does with profiles. At that point it makes sense to link to the profile of the author in the article. Like this: https://abakus.no/articles/488-weekly-uke-44. 

At least I would like a `select` for the author field in the write form. This way we could link them up in the future.

File: packages/core/src/modules/article/article-tag-link-repository.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { type ArticleId, type ArticleTagName } from "@dotkomonline/types"
+import { type Kysely } from "kysely"
+import { type Database } from "@dotkomonline/db"
+
+export interface ArticleTagLinkRepository {
```
Comment (henrikskog) [inline packages/core/src/modules/article/article-tag-link-repository.ts:5]: should these not return the resource that was added/removed?

File: apps/dashboard/src/modules/article/mutations/use-create-article-mutation.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    onMutate: () => {
+      notification.loading({
+        title: "Lager artikkel...",
+        message: "Artikkelen blir opprettet, og du vil bli videresendt til artikkelsiden.",
+      })
+    },
+    onSuccess: (data) => {
+      notification.complete({
+        title: "Artikkel opprettet",
+        message: `Artikkeln "${data.title}" har blitt opprettet.`,
```
Comment (henrikskog) [inline apps/dashboard/src/modules/article/mutations/use-create-article-mutation.ts:19]: Artikkeln

File: apps/dashboard/src/app/(dashboard)/article/write-form.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    label,
+    fields: {
+      title: createTextInput({
+        label: "Tittel",
+        placeholder: "Fadderuka 2023",
+        withAsterisk: true,
+      }),
+      author: createTextInput({
+        label: "Forfattere",
+        placeholder: "Ola Nordmann, Trond-Viggo Torgersen",
```
Comment (henrikskog) [inline apps/dashboard/src/app/(dashboard)/article/write-form.tsx:30]: I would like a select component for this

Comment (henrikskog) [inline apps/dashboard/src/app/(dashboard)/article/write-form.tsx:30]: And then, when we have the user, why not have the ow_user id on the article as well?

File: packages/db/src/migrations/0020_add_article.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { type Kysely, sql } from "kysely"
+import { createTableWithDefaults } from "../utils"
+
+export async function up(db: Kysely<any>) {
+  await createTableWithDefaults("articles", { id: true, createdAt: true, updatedAt: true }, db.schema)
+    .addColumn("title", "text", (col) => col.notNull())
+    .addColumn("author", "text", (col) => col.notNull())
```
Comment (henrikskog) [inline packages/db/src/migrations/0020_add_article.ts:7]: author_ow_user_id?

**Review by henrikskog — APPROVED** [review]
Looks nice! Left some comments

**Review by henrikskog — CHANGES_REQUESTED** [review]
Didn't mean to approve on the last review ;)

File: packages/db/src/migrations/0020_add_article.ts (hunk 2 above)
Comment (junlarsen) [reply packages/db/src/migrations/0020_add_article.ts:9]: I think it's reasonable to say so, I don't think there has been an article on OW without an image for as long as I've been a student

File: packages/db/src/migrations/0020_add_article.ts (hunk 3 above)
Comment (junlarsen) [reply packages/db/src/migrations/0020_add_article.ts:8]: I think in these cases it should be attributed to the image generator, what do you think?

Comment (junlarsen) [reply packages/db/src/migrations/0020_add_article.ts:8]: I think it's probably nice to have a required image alt text field here too

File: packages/db/src/migrations/0020_add_article.ts (hunk 4 above)
Comment (junlarsen) [reply packages/db/src/migrations/0020_add_article.ts:21]: Good catch, agree

File: apps/dashboard/src/app/(dashboard)/article/write-form.tsx (hunk 7 above)
Comment (junlarsen) [reply apps/dashboard/src/app/(dashboard)/article/write-form.tsx:30]: Hard to do with the current state of ow_user would have to search for their ulids right now.

File: packages/db/src/migrations/0020_add_article.ts (hunk 2 above)
Comment (henrikskog) [reply packages/db/src/migrations/0020_add_article.ts:9]: Fair enough!

File: packages/db/src/migrations/0020_add_article.ts (hunk 3 above)
Comment (henrikskog) [reply packages/db/src/migrations/0020_add_article.ts:8]: Yeah it actually seems nice to always have a photographer and alt text

File: packages/core/src/modules/article/article-tag-link-repository.ts (hunk 5 above)
Comment (henrikskog) [reply packages/core/src/modules/article/article-tag-link-repository.ts:5]: Ah yeah true. Okay to leave it out here.

File: apps/dashboard/src/app/(dashboard)/article/write-form.tsx (hunk 7 above)
Comment (henrikskog) [reply apps/dashboard/src/app/(dashboard)/article/write-form.tsx:30]: Should we sync up the attributes of users in cognito with ow_user maybe? It seems very unfortunate to not be able to do stuff like this. However, I agree that it does not need to be a part of this issue, so text input is fine!

Comment (henrikskog) [reply apps/dashboard/src/app/(dashboard)/article/write-form.tsx:30]: I had a bit of a misunderstanding about what was possible to do with ow_user. You can ignore the comment

File: packages/db/src/migrations/0020_add_article.ts (hunk 8 above)
Comment (henrikskog) [reply packages/db/src/migrations/0020_add_article.ts:7]: We don't need this at this moment, you can ignore the comment.

**Review by henrikskog — APPROVED** [review]
You can look over the comments, but I think this is ready to be merged for the purpose of a MVP.

**Comment by junlarsen** [comment]
Oh I completely forgot about this pr, I'll have a closer look later today

**Pull request merged**